src/uwtools/tests/config/test_jinja2.py

```diff
@@ -187,15 +187,55 @@ def test_render(values_file, template, tmp_path):
         assert f.read().strip() == "roses are red, violets are blue"
 
 
-def test_render_dry_run(caplog, values_file, template):
+def test_render_calls__dry_run(template, tmp_path, values_file):
+    outfile = str(tmp_path / "out.txt")
+    with patch.object(jinja2, "_dry_run_template") as dr:
+        render_helper(
+            input_file=template, values_file=values_file, output_file=outfile, dry_run=True
+        )
+        dr.assert_called_once_with("roses are red, violets are blue")
```

Comment:
I could not successfully use `assert_called_once_with(cmd=ANY)`, which is why these are all broken up into separate unit tests. I'd love suggestions for shortening or, better yet, parametrizing them.

Reply:
The `assert_called_once_with()` method (and its ilk -- there's a whole family of related methods) on `Mock` objects are picky about their arguments; in particular, they're literal about whether positional or keyword arguments were used. So, if we've got a function `foo()` and it's mocked as `mock_foo`, `mock_foo.assert_called_once_with(88)` would pass if the code-under-test called `foo(88)`, but not if it called `foo(n=88)`. In the latter case, the assertion would need to be `mock_foo.assert_called_once_with(n=88)`.

For `_dry_run_template()`, the actual call looks like
```
_dry_run_template(template.render())
```
so `dr.assert_called_once_with(ANY)` should pass, since a positional argument is used. If the `_dry_run_template()` call _did_ use a keyword argument, it would be `_dry_run_template(rendered_template=template.render())`, and so the corresponding assertion could be `dr.assert_called_once_with(rendered_template=ANY)`. The `cmd=` keyword argument in your comment probably came from another piece of test code you were looking at, testing a function/method whose argument really was called `cmd`.

Hopefully that helps on that item.

There is some value, though, in being more specific than `ANY` in some cases. `ANY` is great when the argument's actual value truly is something that the test shouldn't know or care about. But even in a case where e.g. it would be good to assert that a specific _type_ of argument is used in the call (like, it matters that it's a `dict` or a `list`), `ANY` might not be strong enough. That seems maybe like the case here.

Maybe something similar to this untested pseudocode could work?
```
@pytest.mark.parametrize("func_arg,func_name,render_extra_kwargs", [
    ("roses are red, violets are blue", "_log_missing_values", {"dry_run": True}), 
    ({"roses_color", "violets_color"}, "_values_needed", {"values_needed": True}),
])
def test_render_makes_expected_calls(func_arg, func_name, render_extra_kwargs, template, tmp_path, values_file):
    outfile = str(tmp_path / "out.txt")
    render_kwargs = {"input_file": template, "values_file": values_file, "output_file": outfile, **render_extra_args}
    with patch.object(jinja2, func_name) as mock:
        render_helper(**render_kwargs)
    mock.assert_called_once_with(func_arg)
```
Something like that might cover 3 of the 4 tests (I only did 2 there). One of them has that special `del` making it asymmetrical with the others, so it might have to stay separate. Or, you could have the `parametrize`d argument also provide a 4th parameter, which would be the name of a key to delete, which would be `None` for 3 of the tests and `"roses_color"` for the other. And the test code could have an `if` condition that only does that config-file mutation for the test that provides a key to delete. Then maybe all 4 tests become symmetrical.

Reply:
How could we include a test for `_write_template` here? It would require either `outfile` or `tmp_path` as an argument, and both would be deemed as undefined variables.

Reply:
Ah, yes. I had to try it to understand, but I see it now. This _could_ be a job for `ANY`, and the following works for me, but I'm not sure if the complexity is worth it. Maybe a few smaller tests with a little repetition are better. You be the judge.
```
@pytest.mark.parametrize(
    "del_key,func_args,func_name,render_extra_kwargs",
    [
        (None, ["roses are red, violets are blue"], "_dry_run_template", {"dry_run": True}),
        ("roses_color", [["roses_color"]], "_log_missing_values", {"dry_run": True}),
        (None, [{"roses_color", "violets_color"}], "_values_needed", {"values_needed": True}),
        (None, [ANY, "roses are red, violets are blue"], "_write_template", {}),
    ],
)
def test_render_makes_expected_calls(
    del_key, func_args, func_name, render_extra_kwargs, template, tmp_path, values_file
):
    if del_key:
        with open(values_file, "r", encoding="utf-8") as f:
            cfgobj = yaml.safe_load(f.read())
        del cfgobj["roses_color"]
        with open(values_file, "w", encoding="utf-8") as f:
            f.write(yaml.dump(cfgobj))
    output_file = str(tmp_path / "out.txt")
    render_kwargs = {
        "input_file": template,
        "output_file": output_file,
        "values_file": values_file,
        **render_extra_kwargs,
    }
    with patch.object(jinja2, func_name) as mock:
        render_helper(**render_kwargs)
    mock.assert_called_once_with(*func_args)
```
I think I might go with smaller, separate tests. This feels like it makes the test too tricky and hard to understand.

Reply:
I'm sorry about that - I need to work on my personal verbose setting. 

Although I like the idea of just having one test to get 4 of the 5 tests needed (the exclusion being `write_template()`'s `stdout` test), I agree that this seems a wee bit too complicated. If I wasn't the developer on this ticket, it would take me a few minutes to figure out everything that's happening here.

Reply:
> I need to work on my personal verbose setting.
No critique implied there -- after I played with the code I understood exactly what you meant, which you accurately expressed. It's difficult to write prose talking about code in an unambiguous way; I try and am sure I fail often.